state-synchronizer/src/tests/unit_tests.rs

```diff
@@ -75,28 +49,59 @@ fn test_remove_requests() {
 
 #[test]
 fn test_request_manager_request_metadata() {
-    let peers = vec![
-        PeerNetworkId::random_validator(),
-        PeerNetworkId::random_validator(),
-    ];
-    let mut request_manager = RequestManager::new(
-        UpstreamConfig::default(),
-        Duration::from_secs(0),
-        Duration::from_secs(30),
-        HashMap::new(),
-    );
-    for peer in peers.iter() {
-        request_manager.enable_peer(peer.clone(), ConnectionOrigin::Outbound);
-    }
+    let (mut request_manager, validators) = generate_request_manager_and_test_validators(0, 2);
+
     assert!(request_manager.get_first_request_time(1).is_none());
 
-    request_manager.add_request(1, vec![peers[0].clone()]);
+    request_manager.add_request(1, vec![validators[0].clone()]);
     request_manager.check_timeout(1);
-    request_manager.add_request(1, vec![peers[1].clone()]);
-    assert!(request_manager.peer_score(&peers[0]).unwrap() < 99.0);
-    assert!(request_manager.peer_score(&peers[1]).unwrap() > 99.0);
+    request_manager.add_request(1, vec![validators[1].clone()]);
+    assert!(request_manager.peer_score(&validators[0]).unwrap() < 99.0);
+    assert!(request_manager.peer_score(&validators[1]).unwrap() > 99.0);
     assert!(
         request_manager.get_first_request_time(1).unwrap()
             <= request_manager.get_last_request_time(1).unwrap()
     );
 }
+
+/// Picks a peer to send a chunk request to (multiple times) and constructs a pick count
+/// for each of the chosen peers.
+fn calculate_pick_counts_for_validators(
+    request_manager: &mut RequestManager,
+    number_of_picks_to_execute: u64,
+) -> HashMap<PeerNetworkId, u64> {
+    let mut pick_counts = HashMap::new();
+
+    for _ in 0..number_of_picks_to_execute {
+        let picked_peers = request_manager.pick_peers();
+        assert_eq!(1, picked_peers.len()); // Ensure only one validator per multicast level
+
+        let picked_peer = picked_peers[0].clone();
+        let counter = pick_counts.entry(picked_peer).or_insert(0);
+        *counter += 1;
+    }
+
+    pick_counts
+}
+
+/// Generates a new request manager with a specified number of validator peers enabled.
+fn generate_request_manager_and_test_validators(
+    request_timeout: u64,
+    num_validators: u64,
+) -> (RequestManager, Vec<PeerNetworkId>) {
+    let mut request_manager = RequestManager::new(
+        UpstreamConfig::default(),
+        Duration::from_secs(request_timeout),
+        Duration::from_secs(30),
+        HashMap::new(),
+    );
+
+    let mut validators = Vec::new();
+    for _ in 0..num_validators {
+        let validator = PeerNetworkId::random_validator();
```

Comment:
Usually for tests you want to remove non-determinism, in this case I don't think it'll introduce flakiness but at least it'll make it easier to debug and should be faster. It's mostly a nit so feel free to ignore this comment, but if you want to fix that you can re-write peer::random as:

```rust
pub fn random(rng: Option<rand::RngCore>) -> Self {
  let buf: [u8; Self::LENGTH] = rng.unwrap_or(rand::OsRng).gen();
  Self(buf)
```

Reply:
Yeah, I don't think this will cause any flakiness in these tests, but good to know for the future. Going to leave things as they are for now and I might return to it down the road.